src/rate_limiter.rs

```diff
@@ -136,16 +145,103 @@ impl RateLimiter {
+        let num_clients = if self.used.contains_key(client_ip) {
+            self.used.len()
+        } else {
+            self.used.len() + 1
+        };
+        cmp::max(MIN_CLIENT_CAPACITY, SOFT_CAPACITY / num_clients as u64)
+    }
+
+    /// Update a client's balance to compensate for initial over-counting.
+    ///
+    /// When a request is made, clients are charged at the maximum size of the data
+    /// being requested. This method compensates the client for the over-counting by
+    /// crediting them the difference between the maximum and the actual size of the response.
+    pub fn deduct_response(
+        &mut self,
+        client_ip: &IpAddr,
+        msg_id: &MessageId,
+        part_count: u32,
+        part_index: u32,
+        payload: &[u8],
+    ) -> Option<u64> {
+        use Response::*;
+
+        // Check that the response isn't a single-part response that we never overcharged for.
+        // This prevents a malicious client from discounting their GET by preceding
+        // it with a PUT with the same `msg_id`.
+        if part_count == 1 && part_index == 0 {
+            match serialisation::deserialise::<UserMessage>(payload) {
+                Ok(UserMessage::Response(response)) => {
+                    match response {
+                        // We overcharged for these, so we let them through.
+                        GetIData { .. } |
+                        GetAccountInfo { .. } |
+                        GetMData { .. } |
+                        GetMDataVersion { .. } |
+                        GetMDataShell { .. } |
+                        ListMDataEntries { .. } |
+                        ListMDataKeys { .. } |
+                        ListMDataValues { .. } |
+                        GetMDataValue { .. } |
+                        ListMDataPermissions { .. } |
+                        ListMDataUserPermissions { .. } |
+                        ListAuthKeysAndVersion { .. } => (),
+                        // These are responses to requests we didn't overcharge for.
+                        // All these responses *should* fit in a single part.
+                        PutIData { .. } |
+                        PutMData { .. } |
+                        MutateMDataEntries { .. } |
+                        SetMDataUserPermissions { .. } |
+                        DelMDataUserPermissions { .. } |
+                        ChangeMDataOwner { .. } |
+                        InsAuthKey { .. } |
+                        DelAuthKey { .. } => return None,
+                    }
+                }
+                _ => return None,
+            }
+        }
+
+        let amount_charged = match self.overcharged.remove(msg_id) {
+            Some(amount) => amount,
+            None => return None,
+        };
+
+        let deduction = amount_charged.saturating_sub(part_count as u64 * MAX_PART_LEN as u64);
+
+        self.used.get_mut(client_ip).map(|used| {
+            *used = used.saturating_sub(deduction);
```

Comment:
we don't have to as the update during next add_message will remove the empty/low_quota entry anyway.

also, to remove in case of `less than the quota` means to calculate the leaked bytes since last update and distribute it among entries, which is actually an `update` call it self.